src/rust/cryptography-x509-validation/src/lib.rs

```diff
@@ -11,6 +11,323 @@ pub mod policy;
+        ))
+    }
+}
+
+pub type Chain<'c> = Vec<Certificate<'c>>;
+type IntermediateChain<'c> = (Chain<'c>, Vec<NameConstraints<'c>>);
+
+pub fn verify<'a, 'chain, B: CryptoOps>(
+    leaf: &'a Certificate<'chain>,
+    intermediates: impl IntoIterator<Item = Certificate<'chain>>,
+    policy: &Policy<'_, B>,
+    store: &'a Store<'chain>,
+) -> Result<Chain<'chain>, ValidationError> {
+    let builder = ChainBuilder::new(Intermediates::new(intermediates, policy)?, policy, store);
+
+    builder.build_chain(leaf)
+}
+
+struct ChainBuilder<'a, 'chain, B: CryptoOps> {
+    intermediates: Intermediates<'chain>,
+    policy: &'a Policy<'a, B>,
+    store: &'a Store<'chain>,
+}
+
+// When applying a name constraint, we need to distinguish between a few different scenarios:
+// * `Applied(true)`: The name constraint is the same type as the SAN and matches.
+// * `Applied(false)`: The name constraint is the same type as the SAN and does not match.
+// * `Skipped`: The name constraint is a different type to the SAN.
+enum ApplyNameConstraintStatus {
+    Applied(bool),
+    Skipped,
+}
+
+impl ApplyNameConstraintStatus {
+    fn is_applied(&self) -> bool {
+        matches!(self, Applied(_))
+    }
+
+    fn is_match(&self) -> bool {
+        matches!(self, Applied(true))
+    }
+}
+
+impl<'a, 'chain, B: CryptoOps> ChainBuilder<'a, 'chain, B> {
+    fn new(
+        intermediates: Intermediates<'chain>,
+        policy: &'a Policy<'a, B>,
+        store: &'a Store<'chain>,
+    ) -> Self {
+        Self {
+            intermediates,
+            policy,
+            store,
+        }
+    }
+
+    fn potential_issuers(
+        &'a self,
+        cert: &'a Certificate<'chain>,
+    ) -> impl Iterator<Item = &'a Certificate<'chain>> + '_ {
+        // TODO: Optimizations:
+        // * Use a backing structure that allows us to search by name
+        //   rather than doing a linear scan
+        // * Search by AKI and other identifiers?
+        self.intermediates
+            .0
+            .iter()
+            .chain(self.store.iter())
+            .filter(|&candidate| candidate.subject() == cert.issuer())
+    }
+
+    fn build_name_constraints(
+        &self,
+        constraints: &mut Vec<NameConstraints<'chain>>,
+        extensions: &Extensions<'chain>,
+    ) -> Result<(), ValidationError> {
+        if let Some(nc) = extensions.get_extension(&NAME_CONSTRAINTS_OID) {
+            let nc: NameConstraints<'chain> = nc.value()?;
+            constraints.push(nc);
+        }
+        Ok(())
+    }
+
+    fn apply_name_constraint(
+        &self,
+        constraint: &GeneralName<'chain>,
+        san: &GeneralName<'_>,
+    ) -> Result<ApplyNameConstraintStatus, ValidationError> {
+        match (constraint, san) {
+            (GeneralName::DNSName(pattern), GeneralName::DNSName(name)) => {
+                if let Some(pattern) = DNSConstraint::new(pattern.0) {
+                    let name = DNSName::new(name.0).unwrap();
```

Comment:
I think we need to handle a malformed `DNSName` here gracefully, rather than upwrapping (but I may have missed something -- maybe we pre-validate this somewhere?).

CC @tetsuo-cpp

Reply:
Opened https://github.com/trailofbits/x509-limbo/pull/102 to exercise this; will push the proper error handling in a moment as well.

Reply:
3060a7031